src/utils/dispatch.test.js

```diff
@@ -0,0 +1,125 @@
+import React, { useReducer, createContext } from 'react'
+import test from 'ava'
+import renderer from 'react-test-renderer'
+
+import {createDispatch, thunkMiddleware} from './dispatch'
+import { async, reject } from 'q'
+
+const middleware = () => {
+	return ({state, dispatch}) => {
+		return next => {
+			return action => {
+				if(action.type === 'TEST'){
+					return dispatch({
+						type: 'MIDDLEWARE',
+						payload: 'success'
+					})
+				}
+				return next(action)
+			}
+		}
+	}
+}
+
+const initState = {
+	test: 'failed',
+	switch: false
+}
+
+const reducer = (state, {type, payload}) => {
+	switch(type){
+		case 'MIDDLEWARE':
+			return {
+				...state,
+				test: payload
+			}
+		case 'DEFAULT':
+			return {
+				...state,
+				test: payload
+			}
+		case 'SWITCH':
+			return {
+				...state,
+				test: payload.test
+			}
+		default:
+			return state
+	}
+}
+
+
+
+test('add dispatch middleware and execute', t => {
+	const Context = createContext()
+
+	function TestComponent(){
+		return (
+			<div></div>
+		)
+	}
+	function Provider(){
+		const [state, reactDispatch] = useReducer(reducer, initState)
+		const dispatch = createDispatch(reactDispatch, state, middleware())
+		return (
+				<Context.Provider value={{state, dispatch}}>
+					<TestComponent dispatch={dispatch} state={state}/>
+				</Context.Provider>
+			)
+	}
+
+	const rendered = renderer.create(<Provider />)
+	const renderedInstance = rendered.root
+
+	const comp = renderedInstance.findByType(TestComponent)
+
+	comp.props.dispatch({type: 'TEST'})
+	t.is(comp.props.state.test, 'success')
+	comp.props.dispatch({type: 'DEFAULT', payload: 'haha'})
+	t.is(comp.props.state.test, 'haha')
+
+
+})
+
+test('test thunkMiddleware', async t => {
+	const Context = createContext()
+
+	const asyncFu = (test) => async ({dispatch, state}) => {
+		await dispatch({
+			type: 'SWITCH',
+			payload:{
+				switch: !state.switch,
+				test
+			}
+		})
+	}
+
+	function TestComponent(){
+		return (
+			<div></div>
+		)
+	}
+	function Provider(){
+		const [state, reactDispatch] = useReducer(reducer, initState)
+		const dispatch = createDispatch(reactDispatch, state, thunkMiddleware())
+		return (
+				<Context.Provider value={{state, dispatch}}>
+					<TestComponent dispatch={dispatch} state={state}/>
+				</Context.Provider>
+			)
+	}
+
+	const rendered = renderer.create(<Provider />)
+	const renderedInstance = rendered.root
+
+	const comp = renderedInstance.findByType(TestComponent)
+	const {state, dispatch} = comp.props
+
+	const switchState = state.switch
+	dispatch(asyncFu('Affenmesserkampf!'))
+	setTimeout(() => {
```

Comment:
Please add comment, why it must explicit exist a timeout for this test.

Reply:
Ok, but what you really want is such as await dispatch(asyncFu());
https://github.com/jeffbski/redux-logic/issues/88
Maybe we should wrap dispatch that include async operations with a global util like process / processHandler that enforce that the return value is a Promise that include as resolve the dispatcher and other return values.

Reply:
I tried it with await, did not get it work